dataforge-distributed/src/jvmMain/kotlin/space/kscience/dataforge/distributed/ServiceWorkspace.kt

```diff
@@ -0,0 +1,119 @@
+    override val targets: Map<String, Meta> = mapOf(),
+) : Workspace, Closeable {
+    private val _port: Int? = port
+
+    override val data: TaskResult<*> = wrapResult(data, Name.EMPTY, Meta.EMPTY)
+
+    override val tasks: Map<Name, Task<*>>
+        get() = context.gather(Task.TYPE)
+
+    private val service = LibService(serviceId, port) {
+        execute of { name, meta, taskRegistry ->
+            if (name == Name.EMPTY) {
+                requireNotNull(dataSerializer) { "Data serializer is not provided on $port" }
+                DataSetPrototype.of(data, dataSerializer)
+            } else {
+                val task = tasks[name] ?: error("Task $name does not exist locally")
+                require(task is SerializableResultTask) { "Result of $name cannot be serialized" }
+                val workspace = ProxyWorkspace(taskRegistry)
+
+                // Local function to capture generic parameter
+                suspend fun <T : Any> execute(task: SerializableResultTask<T>): DataSetPrototype {
+                    val result = task.execute(workspace, name, meta)
+                    return DataSetPrototype.of(result, task.resultSerializer)
+                }
+                execute(task)
+            }
+        }
+    }
+
+    /**
+     * Proxies task calls to right endpoints according to the [TaskRegistry].
+     */
+    private inner class ProxyWorkspace(private val taskRegistry: TaskRegistry) : Workspace by this {
+        override val tasks: Map<Name, Task<*>>
+            get() = object : AbstractMap<Name, Task<*>>() {
+                override val entries: Set<Map.Entry<Name, Task<*>>>
+                    get() = this@ServiceWorkspace.tasks.entries
+
+                override fun get(key: Name): Task<*>? = remoteTask(key) ?: this@ServiceWorkspace.tasks[key]
+            }
+
+        /**
+         * Call default implementation to use [tasks] virtually instead of it in [ServiceWorkspace].
+         */
+        override suspend fun produce(taskName: Name, taskMeta: Meta): TaskResult<*> =
+            super.produce(taskName, taskMeta)
+
+        private fun remoteTask(name: Name): RemoteTask<*>? {
+            val endpoint = taskRegistry.tasks[name] ?: return null
+            val local = this@ServiceWorkspace.tasks[name] ?: error("No task with name $name locally on $port")
+            require(local is SerializableResultTask) { "Task $name result is not serializable" }
+            return RemoteTask(endpoint, local.resultType, local.resultSerializer, local.descriptor, taskRegistry)
+        }
+    }
+
+    /**
+     * Port this workspace is available on.
+     */
+    public val port: Int
+        get() = _port ?: service.port.p
+
+    /**
+     * Start [ServiceWorkspace] as a service.
+     */
+    public fun start(): Unit = service.start()
+
+    /**
+     * Await termination of the service.
+     */
+    public fun awaitTermination(): Unit = service.awaitTermination()
```

Comment:
Why not suspended?

Reply:
Because it is blocking in the google gRPC implementation. It can be made suspend and block separate thread but I do not see any reasons to do so.